### ISSUE TITLE
Support ruamel.yaml 0.17.x

### ISSUE THREAD

**Issue by LourensVeen** [yatiml/yatiml#36]
There's a new release series of ruamel.yaml, 0.17.x, and we should support it. It actually works with the current code, but gives a deprecation warning about the old API.

Unfortunately, the new API seems to be in heavy development, and there are differences between the various 0.17.x versions, and some versions seem to be unable to do what we need at least without ugly hacks.

This needs more investigation, for which I don't have the time right now, so I'm putting it here in hopes of picking it up again in the future.

**Comment by LourensVeen** [comment]
It looks like this is relevant now, see #48. Other packages are apparently moving on to requiring 0.17.x, possibly because they're using the new API, so we're causing compatibility issues.

I've just re-run the test script for all versions since 0.15.71, on Python 3.10 and KUbuntu 22.04. 0.15.x doesn't install anymore, seemingly because there are no wheels for 3.10, and the C code doesn't compile because the Python C bindings have moved on. 

0.16.x mostly works, 0.16.8 doesn't exist, and 0.16.6 gave problems with mypy but passes now, with a newer mypy, or maybe the package was replaced somehow.

0.17.x works, except for 0.17.5 which was released broken.

So, short term, we can extend the dependency version and hobble along for a bit longer. It should probably become `'>=0.15.71,<=0.17.21,!0.17.5'` then.

Longer term, there are still issues with ruamel.yaml:

- It doesn't have a stable API
- The API is changing significantly, and keeping up may become a significant amount of work
- It doesn't support comments very well, and retroactively adding them is proving messy
- It's inherited a poor architecture from PyYAML (components are composed by inheritance), and that architecture negatively affects YAtiML's design
- There are problems with Conda
- It's maintained by a single person who is also the world's last user of SourceForge and Mercurial

Nothing against Anthon (thank you!), who after all maintains ruamel.yaml for free for us all to use and should do whatever  makes him happy, but a bus factor of one is a threat to sustainability and finding more contributors on SourceForge seems unlikely.

The ruamel.yaml package was forked from PyYAML, which at that point was unmaintained and insecure. It seems that PyYAML gets some occasional updates again, and it does have a SafeLoader now. It does still have the architectural deficiencies, which is both a pro and a con because YAtiML uses the ruamel.yaml internals and so porting it to PyYAML may be the least-work option of moving away from ruamel.yaml. There's also a fork of ruamel.yaml called ruyaml, but that seems to have failed to gain much traction so far.

So none of this is fully satisfying. In the I-can-dream category, I would create a new Python YAML library from scratch, with a good architecture and full support for the spec as well as a sane subset of the spec, and contribute it to the standard library. Written in Rust, which I'd finally learn. But well, maybe after I retire.

**Comment by LourensVeen** [comment]
On 0.17.x, we get a
```yatiml/dumper.py:526: PendingDeprecationWarning: 
  dump will be removed, use
  
    yaml=YAML(typ='unsafe', pure=True)
    yaml.dump(...)
  
  instead
    yaml.dump(
```
This is a pretty major issue. This `YAML` class seems to no longer be using a `Dumper`, and YAtiML uses a custom Dumper. So we'd have to figure out how to port YAtiML to the new way of doing things of ruamel.yaml, while that new way is evolving. Also it's not easy to understand, undocumented, and there are no guarantees that any of it will not change from version to version.

So maybe that new YAML library doesn't sound so bad. The only problem is that we'd have to effectively rebuild YAtiML on top of it, since it's so closely integrated with the ruamel.yaml (and PyYAML) architecture.

So what to do? Maybe in increasing order of amount of work and also of result satisfaction:

1. Extend the version range so that people can work, albeit with that warning
2. Investigate porting to PyYAML, which being barely maintained can be expected to be more stable
3. Create a new YAML library for Python and either integrate YAtiML or something like it into it, or rebuild YAtiML on top of it.

How hard can it be to write a PEG grammer for YAML? Ah, almost impossible: https://github.com/yaml/yaml-grammar . So there's a

4. Create a new YAML-like but cleaner markup language, create a new library for it for Python, then either integrate YAtiML or something like it into it, or rebuild YAtiML on top of it. Maybe next weekend?

**Comment by LourensVeen** [comment]
Released in 0.11.0.
